File: artifacts/api-server/src/routes/islamic.ts
import { Router, type IRouter } from "express";
import { eq, and, or, isNull } from "drizzle-orm";
import { db, quranProgressTable, islamicActivitiesTable, activityLogsTable } from "@workspace/db";
import { z } from "zod";

const router: IRouter = Router();

// ─── Schemas ──────────────────────────────────────────────────────────────────

const InitQuranProgressBody = z.object({
  targetDate: z.string().optional().nullable(),
  currentSurah: z.number().int().min(1).max(114).optional(),
  currentPage: z.number().int().min(1).max(604).optional(),
  dailyTarget: z.number().int().min(1).optional(),
  notes: z.string().optional().nullable(),
});

const UpdateQuranProgressBody = z.object({
  currentSurah: z.number().int().min(1).max(114).optional(),
  currentPage: z.number().int().min(1).max(604).optional(),
  targetDate: z.string().optional().nullable(),
  dailyTarget: z.number().int().min(1).optional(),
  notes: z.string().optional().nullable(),
});

const LogDeedBody = z.object({
  status: z.enum(["completed", "intended"]).default("completed"),
  date: z.string().optional(),
  hijriDate: z.string().optional().nullable(),
  notes: z.string().optional().nullable(),
});

// ─── Quran Progress ───────────────────────────────────────────────────────────

router.get("/quran", async (_req, res): Promise<void> => {
  const [progress] = await db
    .select()
    .from(quranProgressTable)
    .orderBy(quranProgressTable.createdAt)
    .limit(1);

  if (!progress) {
    res.status(404).json({ error: "No Quran progress found" });
    return;
  }

  res.json(serializeQuran(progress));
});

router.post("/quran", async (req, res): Promise<void> => {
  const parsed = InitQuranProgressBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  // Calculate daily target if targetDate provided
  let dailyTarget = parsed.data.dailyTarget ?? 2;
  const currentPage = parsed.data.currentPage ?? 1;
  if (parsed.data.targetDate) {
    const daysLeft = Math.max(
      1,
      Math.ceil((new Date(parsed.data.targetDate).getTime() - Date.now()) / 86400000)
    );
    const pagesLeft = 604 - currentPage;
    dailyTarget = Math.max(1, Math.ceil(pagesLeft / daysLeft));
  }

  const [progress] = await db
    .insert(quranProgressTable)
    .values({
      targetDate: parsed.data.targetDate ?? null,
      currentSurah: parsed.data.currentSurah ?? 1,
      currentPage,
      dailyTarget,
      notes: parsed.data.notes ?? null,
    })
    .returning();

  res.status(201).json(serializeQuran(progress));
});

router.patch("/quran", async (req, res): Promise<void> => {
  const parsed = UpdateQuranProgressBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const [existing] = await db.select().from(quranProgressTable).limit(1);
  if (!existing) {
    res.status(404).json({ error: "No Quran progress found. Initialize first." });
    return;
  }

  const update: Record<string, unknown> = {};
  if (parsed.data.currentPage !== undefined) update.currentPage = parsed.data.currentPage;
  if (parsed.data.currentSurah !== undefined) update.currentSurah = parsed.data.currentSurah;
  if (parsed.data.targetDate !== undefined) update.targetDate = parsed.data.targetDate;
  if (parsed.data.dailyTarget !== undefined) update.dailyTarget = parsed.data.dailyTarget;
  if (parsed.data.notes !== undefined) update.notes = parsed.data.notes;

  // Recalculate daily target if targetDate changed
  if (parsed.data.targetDate && !parsed.data.dailyTarget) {
    const currentPage = (parsed.data.currentPage ?? existing.currentPage);
    const daysLeft = Math.max(
      1,
      Math.ceil((new Date(parsed.data.targetDate).getTime() - Date.now()) / 86400000)
    );
    const pagesLeft = 604 - currentPage;
    update.dailyTarget = Math.max(1, Math.ceil(pagesLeft / daysLeft));
  }

  const [updated] = await db
    .update(quranProgressTable)
    .set(update)
    .where(eq(quranProgressTable.id, existing.id))
    .returning();

  res.json(serializeQuran(updated));
});

// ─── Deeds Catalog ────────────────────────────────────────────────────────────

router.get("/deeds", async (req, res): Promise<void> => {
  const todayOnly = req.query.todayOnly === "true";
  const date = typeof req.query.date === "string" ? req.query.date : new Date().toISOString().split("T")[0];

  const dayOfWeek = new Date(date + "T12:00:00").getDay(); // 0=Sun … 5=Fri

  const deeds = await db
    .select()
    .from(islamicActivitiesTable)
    .where(eq(islamicActivitiesTable.isActive, true));

  // Filter for today's deeds: always-available (no day_of_week constraint) + today's weekday
  const filtered = todayOnly
    ? deeds.filter((d) => d.dayOfWeek === null || d.dayOfWeek === dayOfWeek)
    : deeds;

  filtered.sort((a, b) => a.sortOrder - b.sortOrder);
  res.json(filtered.map(serializeDeed));
});

router.post("/deeds/:id/log", async (req, res): Promise<void> => {
  const { id } = req.params;
  const parsed = LogDeedBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const date = parsed.data.date ?? new Date().toISOString().split("T")[0];

  const [log] = await db
    .insert(activityLogsTable)
    .values({
      activityId: id,
      status: parsed.data.status,
      date,
      hijriDate: parsed.data.hijriDate ?? null,
      notes: parsed.data.notes ?? null,
    })
    .returning();

  res.status(201).json(serializeLog(log));
});

router.get("/deeds/logs", async (req, res): Promise<void> => {
  const date = typeof req.query.date === "string"
    ? req.query.date
    : new Date().toISOString().split("T")[0];

  const logs = await db
    .select({
      id: activityLogsTable.id,
      activityId: activityLogsTable.activityId,
      status: activityLogsTable.status,
      date: activityLogsTable.date,
      hijriDate: activityLogsTable.hijriDate,
      notes: activityLogsTable.notes,
      loggedAt: activityLogsTable.loggedAt,
      activityName: islamicActivitiesTable.name,
      activityCategory: islamicActivitiesTable.category,
      activityRewardText: islamicActivitiesTable.rewardText,
    })
    .from(activityLogsTable)
    .innerJoin(islamicActivitiesTable, eq(activityLogsTable.activityId, islamicActivitiesTable.id))
    .where(eq(activityLogsTable.date, date));

  res.json(logs.map((l) => ({
    ...l,
    loggedAt: l.loggedAt.toISOString(),
  })));
});

// ─── Serializers ──────────────────────────────────────────────────────────────

function serializeQuran(p: typeof quranProgressTable.$inferSelect) {
  const pagesLeft = p.totalPages - p.currentPage;
  const percentComplete = Math.round((p.currentPage / p.totalPages) * 100);
  let daysToComplete: number | null = null;
  if (p.targetDate) {
    daysToComplete = Math.max(0, Math.ceil((new Date(p.targetDate).getTime() - Date.now()) / 86400000));
  }
  return {
    ...p,
    pagesLeft,
    percentComplete,
    daysToComplete,
    updatedAt: p.updatedAt.toISOString(),
    createdAt: p.createdAt.toISOString(),
  };
}

function serializeDeed(d: typeof islamicActivitiesTable.$inferSelect) {
  return { ...d, createdAt: d.createdAt.toISOString() };
}

function serializeLog(l: typeof activityLogsTable.$inferSelect) {
  return { ...l, loggedAt: l.loggedAt.toISOString() };
}

export default router;

File: artifacts/api-server/routes/islamic.py
from __future__ import annotations

from datetime import datetime, date as date_cls
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from sqlmodel import Session, select, and_
from database import get_session
from models import Khatmah, IslamicActivity, ActivityLog
from utils import today_str
from clerk_auth import get_current_user_id

router = APIRouter()

_TYPE_TO_CATEGORY = {
    "fard":     "prayer",
    "sunnah":   "sunnah",
    "mostahab": "other",
}

_LOG_STATUS_TO   = {"completed": "Completed", "intended": "Intended"}
_LOG_STATUS_FROM = {"Completed": "completed", "Intended": "intended"}


# ─── Serializers ─────────────────────────────────────────────────────────────
def serialize_khatmah(k: Khatmah) -> dict:
    pages_left = k.total_pages - k.current_page
    percent_complete = round((k.current_page / k.total_pages) * 100) if k.total_pages else 0
    days_to_complete = None
    daily_target = 2
    if k.target_date:
        target = date_cls.fromisoformat(k.target_date)
        days_to_complete = max(0, (target - date_cls.today()).days)
        if days_to_complete > 0:
            daily_target = max(1, -(-pages_left // days_to_complete))
    return {
        "id": k.id,
        "name": k.name,
        "type": k.type,
        "status": k.status,
        "currentSurah": 1,
        "currentPage": k.current_page,
        "totalPages": k.total_pages,
        "targetDate": k.target_date,
        "dailyTarget": daily_target,
        "pagesLeft": pages_left,
        "percentComplete": percent_complete,
        "daysToComplete": days_to_complete,
        "notes": None,
        "updatedAt": k.created_at.isoformat(),
        "createdAt": k.created_at.isoformat(),
    }


def serialize_deed(a: IslamicActivity) -> dict:
    return {
        "id": a.id,
        "name": a.name,
        "arabicName": None,
        "rewardText": a.reward_text or "",
        "category": _TYPE_TO_CATEGORY.get(a.type, "other"),
        "hijriMonth": a.hijri_month,
        "hijriDay": a.hijri_day,
        "dayOfWeek": None,
        "isActive": not a.is_archived,
        "sortOrder": 0,
        "type": a.type,
        "createdAt": a.created_at.isoformat(),
    }


def serialize_log(lg: ActivityLog) -> dict:
    return {
        "id": lg.id,
        "activityId": lg.activity_id,
        "status": _LOG_STATUS_FROM.get(lg.status, "completed"),
        "date": lg.logged_at.strftime("%Y-%m-%d"),
        "hijriDate": lg.hijri_date,
        "notes": None,
        "loggedAt": lg.logged_at.isoformat(),
    }


# ─── Request bodies ───────────────────────────────────────────────────────────
class InitQuranBody(BaseModel):
    name: Optional[str] = "My Khatmah"
    targetDate: Optional[str] = None
    currentSurah: Optional[int] = None
    currentPage: Optional[int] = None
    dailyTarget: Optional[int] = None
    notes: Optional[str] = None


class UpdateQuranBody(BaseModel):
    currentSurah: Optional[int] = None
    currentPage: Optional[int] = None
    targetDate: Optional[str] = None
    dailyTarget: Optional[int] = None
    notes: Optional[str] = None


class LogDeedBody(BaseModel):
    status: Optional[str] = "completed"
    date: Optional[str] = None
    hijriDate: Optional[str] = None
    notes: Optional[str] = None


# ─── Khatmah (Quran progress) routes ─────────────────────────────────────────
@router.get("/quran")
def get_quran(
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    k = session.exec(
        select(Khatmah).where(Khatmah.user_id == user_id, Khatmah.is_archived == False).order_by(Khatmah.created_at)
    ).first()
    if not k:
        raise HTTPException(status_code=404, detail="No Khatmah found")
    return JSONResponse(content=serialize_khatmah(k))


@router.post("/quran", status_code=201)
def init_quran(
    body: InitQuranBody,
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    current_page = body.currentPage or 1
    k = Khatmah(
        user_id=user_id,
        name=body.name or "My Khatmah",
        target_date=body.targetDate,
        current_page=current_page,
    )
    session.add(k)
    session.commit()
    session.refresh(k)
    return JSONResponse(content=serialize_khatmah(k), status_code=201)


@router.patch("/quran")
def update_quran(
    body: UpdateQuranBody,
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    k = session.exec(
        select(Khatmah).where(Khatmah.user_id == user_id, Khatmah.is_archived == False)
    ).first()
    if not k:
        raise HTTPException(status_code=404, detail="No Khatmah found. Initialize first.")

    if body.currentPage is not None:
        k.current_page = body.currentPage
    if body.targetDate is not None:
        k.target_date = body.targetDate

    session.add(k)
    session.commit()
    session.refresh(k)
    return JSONResponse(content=serialize_khatmah(k))


# ─── Deeds routes ─────────────────────────────────────────────────────────────
@router.get("/deeds")
def list_deeds(
    todayOnly: Optional[str] = None,
    date: Optional[str] = None,
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    deeds = session.exec(
        select(IslamicActivity).where(IslamicActivity.is_archived == False)
    ).all()

    if todayOnly == "true":
        from utils import get_hijri_day
        target_date = date or today_str()
        hijri_day = get_hijri_day(target_date)
        deeds = [d for d in deeds if d.hijri_day is None or d.hijri_day == hijri_day]

    return JSONResponse(content=[serialize_deed(d) for d in deeds])


@router.post("/deeds/{deed_id}/log", status_code=201)
def log_deed(
    deed_id: str,
    body: LogDeedBody,
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    stored_status = _LOG_STATUS_TO.get(body.status or "completed", "Completed")
    log_date = body.date or today_str()
    lg = ActivityLog(
        user_id=user_id,
        activity_id=deed_id,
        status=stored_status,
        hijri_date=body.hijriDate,
        logged_at=datetime.fromisoformat(log_date + "T12:00:00"),
    )
    session.add(lg)
    session.commit()
    session.refresh(lg)
    return JSONResponse(content=serialize_log(lg), status_code=201)


@router.get("/deeds/logs")
def get_deed_logs(
    date: Optional[str] = None,
    user_id: str = Depends(get_current_user_id),
    session: Session = Depends(get_session),
):
    target_date = date or today_str()
    start = datetime.fromisoformat(target_date + "T00:00:00")
    end = datetime.fromisoformat(target_date + "T23:59:59")
    logs = session.exec(
        select(ActivityLog).where(
            and_(ActivityLog.user_id == user_id, ActivityLog.logged_at >= start, ActivityLog.logged_at <= end)
        )
    ).all()

    activity_ids = list({lg.activity_id for lg in logs})
    activities = {}
    if activity_ids:
        activities = {a.id: a for a in session.exec(
            select(IslamicActivity).where(IslamicActivity.id.in_(activity_ids))
        ).all()}

    result = []
    for lg in logs:
        d = serialize_log(lg)
        act = activities.get(lg.activity_id)
        if act:
            d["activityName"] = act.name
            d["activityCategory"] = _TYPE_TO_CATEGORY.get(act.type, "other")
            d["activityRewardText"] = act.reward_text or ""
        result.append(d)

    return JSONResponse(content=result)
